robosat/unet.py

```diff
@@ -1,100 +1,114 @@
-'''The "U-Net" architecture for semantic segmentation.
+'''U-Net inspired encoder-decoder architecture with a ResNet encoder as proposed by Alexander Buslaev.
 
 See:
 - https://arxiv.org/abs/1505.04597 - U-Net: Convolutional Networks for Biomedical Image Segmentation
 - https://arxiv.org/abs/1411.4038  - Fully Convolutional Networks for Semantic Segmentation
+- https://arxiv.org/abs/1512.03385 - Deep Residual Learning for Image Recognition
+- https://arxiv.org/abs/1801.05746 - TernausNet: U-Net with VGG11 Encoder Pre-Trained on ImageNet for Image Segmentation
+- https://arxiv.org/abs/1806.00844 - TernausNetV2: Fully Convolutional Network for Instance Segmentation
 
 '''
 
 import torch
 import torch.nn as nn
 
+from torchvision.models import resnet50
```

Comment:
Should this be part of the model configurations? I mean, would it help if users saw that by default Robosat would use a `Resnet50` and they could choose other configurations available in torchvision if they want to?

- https://pytorch.org/docs/master/torchvision/models.html#id3

Reply:
Maybe. At the moment we also still have the model name in the config

https://github.com/mapbox/robosat/blob/b2b6df26dd091da9344982d440c536b90ec30e4f/config/model-unet.toml#L8-L9

but don't use it at all and simply always use our U-Net adaption. The PSPNet in multi-class mode will not work with how we currently serialize and deserialize segmentation probabilities, too.

I think we need to adapt a bit more code than just swapping out the ResNets:

https://github.com/mapbox/robosat/blob/5a1479c5223046768623a4649a62e1b61c381dff/robosat/unet.py#L94-L108

What we could do is abstract away the feature extractor, then our UNet class would take such a FeatureExtractor and get feature maps of specific resolutions and channels from it, and construct a corresponding decoder. I don't think it's worth it, though.